### ISSUE TITLE
Add `exceptWithoutPropertyFallback: []` to `function-allowed-list`

### ISSUE THREAD

**Issue by AlexanderShushunov** [stylelint/stylelint#8291]
### What is the problem you're trying to solve?

The `function-allowed-list` rule is a powerful tool for ensuring that only CSS functions supported by our target browsers are used. For example, if we aim to support Chrome 47, we might want to block the use of the `min()` function, which is supported starting from Chrome 79.

However, in many cases, developers can provide fallbacks to ensure compatibility with older browsers while enhancing the user experience for modern ones. For instance:

```css
  width: 200px; /* Fallback for older browsers */
  width: min(50%, 300px); /* Advanced functionality for modern browsers */
```

Currently, `function-allowed-list` prohibits this usage even when a fallback is present. This limits developers from leveraging modern CSS capabilities responsibly.

### What solution would you like to see?

Allow the use of restricted functions if a valid fallback is provided

**Comment by romainmenke** [comment]
Hi @AlexanderShushunov,

Thank you for opening an issue about this.

[Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcICGACYAdAdilB3AllAC4AW8KATAAxUAOAHgNwoD0AVCgGJIA2PARkjABrFADMIAJxQQesaf0kRcAZxiSVKNiyw4CxMigC2+TAAoArFQCkAGhQBmGgwCUzdigCCUAG5JMYDBQ4gCuAUT4EJi8+EQAnuJSxtDq2IrKahpaOpgAvlhY-Gi6eISk5NR0TKwc3HyCIonSsvIo6arqmtol+uXGppY29k5VbjUoKvE8MDymRAC0UPgqSPzT87OYMKHhkZjzvDzKQRvLRCjz8yi0SgDmkjAqKvg+2zCYJP6BRu-n3XlYEC2EBifDTABySB+CBAMHoUNo0wAdGAnkDwFFQbcYRhsCh0LD6ER3lAVATyABtEo4AmTOLTTYLSCYLHzSb+KBISRQAklAC6thKBMkIWmZIQxTxNJBYTAESiBz4xygp0m5JQFL5JXyeXRzKxnCkRiQRBhACsVFF0bBaCocUKQHSGXN1QSeCbHkQCYK8bSpjM5vN9fhbmyiByuTyJW6PWqQFhciBckA)

Thank you also for providing a PR with a proposed fix.
We normally first discuss the issue and then mark it as "ready to implement".
Let's also hear from others about this :)

Your proposed solution doesn't detect the use of `@supports` while also not taking custom props into consideration. I think this shows the challenge in providing a good solution.


`@supports`:

```css
.foo {
  width: 200px; /* Fallback for older browsers */
}

@supprts (width: min(1, 2)) {
  .foo {
    width: min(50%, 300px); /* Advanced functionality for modern browsers */
  }
}
```

custom props without `@supports`:

```css
:root {
  --x-size: 200px; /* will never be used, even in older browsers */
  --x-size: min(50%, 300px); /* will always be used even in browsers that do not support it, leading to issues */
}
```

custom props with `@supports`:

```css
:root {
  --x-size: 200px; /* Fallback for older browsers */
}

@supprts (width: min(1, 2)) {
  :root {
    --x-size: min(50%, 300px); /* Advanced functionality for modern browsers */
  }
}
```

-----

We also should not assume why authors are only allowing specific functions.
Enabling this change without a rule config might be breaking for some users.

-----

I am not sure what the right pattern is for this specific issue.
I can imagine that this same issue exists for multiple rules?

Maybe a new rule to require progressive enhancement patterns could also be an option?
Such a new rule could specialize in detecting the absence of progressive enhancements.

**Comment by AlexanderShushunov** [comment]
Hi @romainmenke,

Thank you for your feedback.

Can we add a secondary option to the rule? It allows specifying the rule behavior only when necessary, so we will not break any user experience.

E.g.
```json
 "function-allowed-list": [
      ["min", "max"],
      {
        "allowWithFallback": ["min"]
      }
    ]
```
or 
```json
 "function-allowed-list": [
      ["min", "max"],
      {
        "allowWithFallback": true
      }
    ]
```

**Comment by Mouvedia** [comment]
[`declaration-block-no-duplicate-properties`](https://stylelint.io/user-guide/rules/declaration-block-no-duplicate-properties#ignore-consecutive-duplicates) rule has `ignore: ["consecutive-duplicates"]` and `ignore: ["consecutive-duplicates-with-different-values"]`; I think we could consider adding `ignore: ["consecutive-duplicate-properties"]` to `function-allowed-list` and `function-disallowed-list` rules.

I don't have a strong opinion on this.

**Comment by AlexanderShushunov** [comment]
@Mouvedia, as @romainmenke mentioned, it would be better to include an additional check for whether a fallback is provided within an `@supports` rule. Therefore, `consecutive-duplicate-properties` might not be suitable for addressing this specific case.

**Comment by AlexanderShushunov** [comment]
@romainmenke, I believe it would be better not to create a new rule. For scenarios where functions are allowed or disallowed based on browser support, it's more practical to adjust the existing rule in a single place.

As for using `stylelint-disable-line function-allowed-list` comments, this approach can be quite annoying to apply repeatedly—it's actually one of the reasons I opened this issue ;). 
Additionally, this solution is somewhat fragile. If a fallback is accidentally removed, it can result in unexpected behavior without triggering any linting errors.

**Comment by romainmenke** [comment]
> I believe it would be better not to create a new rule. For scenarios where functions are allowed or disallowed based on browser support, it's more practical to adjust the existing rule in a single place.

I am not sure if `*-allowed-list` rules are typically used to ensure browser support and even if they are sometimes used for that, if these are the best method.

I think another solution could be to leverage the proposed [`languageOptions`](https://github.com/stylelint/stylelint/issues/8154)

This would allow you to describe your support targets and the syntax that exists in those targets.

[`function-no-unknown`](https://stylelint.io/user-guide/rules/function-no-unknown) for example could use that information to consider `min` as unknown following your config.

In turn not warning when there are fallbacks/progressive enhancements would be nice there.

-----

TL;DR; I think there are multiple approaches to this issue:
- extending the rules your are currently using with new behavior
- using different rules and optionally extending their behavior
- new rules

We should pick what is best overal :)

**Comment by romainmenke** [comment]
`allowAsProgressiveEnhancement ` might be a better name? (or a variation thereof)

```js
 "function-allowed-list": [
      ["min", "max"],
      {
        "allowAsProgressiveEnhancement": ["min"]
      }
    ]
```

This covers any pattern used to write more modern CSS in a codebase targeting older browser versions. (multiple declarations for the same property, `@supports`, ...)

**Comment by AlexanderShushunov** [comment]
@romainmenke, I believe "extending the rules" is the best option here, as it avoids increasing the complexity of the configuration. By simply adding an optional parameter, we can gradually fine-tune the behavior without introducing new rules.

Regarding the `allowAsProgressiveEnhancement` parameter, I think it’s a great name. Should we also consider adding it to the `function-disallowed-list` rule for consistency?

**Comment by jeddy3** [comment]
Out of @romainmenke's three options, I think there's value in exploring new rules specifically for progressive enhancement. Something that ties into Baseline. For example,

```json
{
  "rules":
    "function-no-above-baseline": ["widely-available", {
      "ignore": [
        "in-at-support",
        "after-duplicate-property"
      ]
   }]
}
```

(Alternatively, we could have `true` for the primary option and tie into a baseline set in `languageOptions`. Need bike-shed the rule names, too.)

That'd give us:

- `*-no-unknown` - detect anything unknown in the specs
- `*-no-deprecated` - detect anything deprecated in the specs
- `*-no-above-baseline` - detect anything above baseline
- `*-disallowed-list` - disallow anything (that is known in the specs, not deprecated and within baseline)

With the following already caught by [`declaration-block-no-duplicate-custom-properties`](https://stylelint.io/user-guide/rules/declaration-block-no-duplicate-custom-properties), I think a new suite of `*-no-above-baseline` rules could round out our coverage of use cases:

```css
:root {
  --x-size: 200px; /* will never be used, even in older browsers */
  --x-size: min(50%, 300px); /* will always be used even in browsers that do not support it, leading to issues */
}
```

**Comment by ryo-manba** [comment]
I agree with [romain's opinion](https://github.com/stylelint/stylelint/issues/8291#issuecomment-2584431895). 

I think this request should be considered separately from Baseline. 
Since Baseline is still a relatively new concept, as Alexander mentioned, some projects likely manage their own lists of supported browsers.

**Comment by jeddy3** [comment]
> I think this request should be considered separately from Baseline.

That sounds good to me. It aligns with our recently changed vision of providing users with more customisation, including overlapping rules. 

We can add secondary options to our `*-(dis)allowed-list` rules _and_ add rules dedicated to baseline (https://github.com/stylelint/stylelint/issues/8391).

We have many options spread over our rules, so we have a [set of conventions](https://stylelint.io/developer-guide/rules#write-the-rule) to ensure a consistent experience, typically:

- rules are strict by default and made more permissive with the `ignore` and `ignore*` options
- options target specific constructs 

With those in mind, two possible approaches could be:

```json
{
  "rules": {
    "function-disallowed-list": [["min"], {
      "ignore": [
        "inside-at-supports",
        "after-duplicate-property"
      ]
   }]
  }
}
```

Or, for more fine-grained control:


```json
{
  "rules": {
    "function-disallowed-list": [["min"], {
      "ignoreFunctionsInsideAtSupports": ["min"],
      "ignoreFunctionsAfterDuplicateProperty": ["min"]
    }]
  }
}
```

We can add additional options for other patterns used for progressive enhancement. As with https://github.com/stylelint/stylelint/issues/8391, options are specific to the targeted construct, e.g. `after-duplicate-property` wouldn't be appropriate for `selector-disallowed-list`.

**Comment by Mouvedia** [comment]
I prefer ’"consecutive-duplicate-properties"’ over  ’"after-duplicate-properties"’ because the property which is considered the duplicate is the following one in the former but for the latter the duplicate precede it which is illogical to me.

**Comment by jeddy3** [comment]
> I prefer `"consecutive-duplicate-properties"`

SGTM as it [matches our existing keywords](https://stylelint.io/user-guide/rules/declaration-block-no-duplicate-properties/#optional-secondary-options), too.

**Comment by romainmenke** [comment]
Also see [`first-valid()`](https://drafts.csswg.org/css-values-5/#first-valid)

This will also be a way to write the CSS from the original post:

```css
.foo {
  width: first-valid(min(50%, 300px); 200px);
}
```

The name of `first-valid()` is likely to change but the feature is definitely going to ship at some point.

That implies that each plugin will need three options to cover all methods of progressive enhancements.

I am unsure if users will ever only set one or two and not the others?
I personally prefer a single option (i.e. `allowAsProgressiveEnhancement`) but I also understand that options normally target specific constructs.

------

`consecutive-duplicate-properties` might be too simplistic:
- it doesn't need to consecutive, interleaving other declarations is fine
- a preceding declaration with the same property might have the same value
- the declaration itself might only be invalid at computed value time (IACVT), in which case no preceding declaration will ever be applied

So these are all possible and wouldn't fit well with this name:

```css
.foo {
  width: 200px;
  height: 50px;
  width: min(50%, 300px); /* should not error */
}
```

```css
.foo {
  width: min(50%, 300px); /* should error */
  width: min(50%, 300px); /* should error */
}
```

```css
.foo {
  width: 200px;
  width: min(var(--some-percentage), 300px); /* should error */
}
```

**Comment by jeddy3** [comment]
All good points, thanks @romainmenke.

> Also see [first-valid()](https://drafts.csswg.org/css-values-5/#first-valid)

Fascinating. I did wonder if something like that was in the pipeline after looking through the mixin spec the other day.

> I am unsure if users will ever only set one or two and not the others?

I think that's the crux of the issue. I'm not sure either, but we've generally found (through the various requests for rules and options over the years) that our users who use these types of rules want to enforce strict CSS conventions. 

Our rule conventions, e.g. targeting specific constructs, lean into this desire for granularity. 

`first-valid()` is a good case in point. When it lands, I can foresee users wanting to enforce this _specific method_ of progressive enhancement over another, e.g. allow this method:

```css
.foo {
  width: first-valid(min(50%, 300px); 200px);
}
```

but disallow this (and other methods):

```css
.foo {
  width: 200px;
  width: min(50%, 300px); 
}
```

I assume the same may be true now, e.g. wanting to enforce the `@supports` method over a property fallback method or vice versa (for specific properties). 

If we want to support that, it looks like we have two approaches:

- add a broad `ignoreFunctionsAsProgressiveEnhancement: []` / `ignore: ["as-progressive-enhancement"]` option to the `(dis)allow-list` rules, and then add other rules to enforce specific methods of progressive enhancement
- add granular options to the `(dis)allow-list` rules that let users enforce specific methods of progressive enhancement

In both instances, the rules and the options would be specific to the construct. For example, for `selector-disallowed-list`, we may only need to account for the `@supports` method:

```json
{
  "rules": {
    "selector-disallowed-list": [[":has()"], {
      "ignore": ["inside-at-supports"]
    }]
  }
}
```

```css
@supports selector(:has())) {
  .foo:has(a) {}
}
```

I don't have a strong preference, but my gut says granular options, as this approach has served our users well so far.

Whichever approach we take, we'd follow our usual process of only adding requested features, e.g. we'd start with either the granular approach to the specific `function-allowed-list` rule use case:

```json
{
  "rules": {
    "function-allowed-list": [["min"], {
      "except": [
        "without-property-fallback"
      ]
   }]
  }
}
```

Or the broad one (which would, for now, only check the property fallback method behind the scenes):

```json
{
  "rules": {
    "function-allowed-list": [["min"], {
      "except": [
        "without-progress-enhancement"
      ]
   }]
  }
}
```

(["except" inverts the primary option](https://stylelint.io/developer-guide/rules#keyword-ignore-and-except). Option names to be bike-shedded, and thank you for your insightful comments about that, especially the IACVT consideration.)

**Comment by romainmenke** [comment]
> add granular options

Sounds good to me.
This indeed gives more capabilities to users :)

**Comment by AlexanderShushunov** [comment]
Apologies for the silence — the discussion slipped off my radar.
Thank you, @ryo-manba, for implementing the solution!
